JavaScript/chapter02/p05_sum_lists/avc278.js

```diff
@@ -0,0 +1,105 @@
+// Sum Lists: You have two numbers represented by a linked list, where each node contains a single digit. The digits are
+// stored in reverse order, such that the 1's digit is at the head of the list. Write a function that adds the two
+// numbers and returns the sum as a linked list.
+// Example
+// Input: (7-> 1 -> 6) + (5 -> 9 -> 2). That is, 617 + 295.
+// Output: 2 -> 1 -> 9. That is, 912.
+//
+// Follow up: Suppose the digits are stored in forward order. Repeat the above problem.
+// Example
+// Input: (6 -> 1 -> 7) + (2 -> 9 -> 5). That is, 617 + 295.
+// Output: 9 -> 1 -> 2. That is, 912.
+
+const assert = require("assert");
+const {
+  arrayToLinkedList,
+  compareLinkedLists,
+  LinkedListNode,
+} = require("../../lib/avc278/linkedlist");
+
+/**
+ * Partitions a linked list based on the input partition, with lesser values to the left and greater values to the right
+ * @param   {LinkedListNode} listOne input linked list 1 to be added
+ * @param   {LinkedListNode} listTwo input linked list 2 to be added
+ * @return  {LinkedListNode}         ouput sum of input linked lists
+ *
+ * For this problem, we need to keep a carry in case two node values, when summed, are greater than 10. We also need to
+ * keep track of the sum linked list head, as well as the previous and current nodes of the sum linked list.
+ * We traverse through our two linked lists, adding their node values together, storing them into a temp sum, then
+ * creating a new node based on that sum value, noting whether the sum is greater than 10 or not, and adding it on to
+ * the sumHead linked list. Traversing through listOne requires a runtime of O(N) where N is the length of the list,
+ * and similarly, traversing through listTwo requires a runtime of O(M) where M is the length of the list. Since they
+ * travel at the same pace, this will require a runtime of of max(M, N). Since we need to construct an answer based on
+ * the sum of the two lists, we will construct a linked list whose length is the maximum of M and N, plus 1, which
+ * consequently also requires max(M, N) space.
+ * Runtime: O(max(M, N))
+ * Space:   O(max(M, N))
+ *
+ */
+const sumLists = (listOne, listTwo) => {
+  let carry = 0;
+  let sumHead;
+  let prevNode;
+  let currNode;
+  while (carry > 0 || listOne !== null || listTwo !== null) {
+    let sum = carry;
+    if (listOne !== null) {
+      sum += listOne.val;
+      listOne = listOne.next;
+    }
+    if (listTwo !== null) {
+      sum += listTwo.val;
+      listTwo = listTwo.next;
+    }
+
+    carry = 0;
+    if (sum >= 10) {
+      carry = 1;
+    }
+    currNode = new LinkedListNode(sum % 10);
```

Comment:
nit: it's probably better to declare this here: https://wiki.c2.com/?DeclareVariablesAtFirstUse

that way readers know that the three variables at the top are expected to survive beyond the loop / across loop executions, but `currNode` does not need to.